### ISSUE TITLE
Performance issues in `deepMap`

### ISSUE THREAD

**Issue by josdejong** [josdejong/mathjs#3262]
After extending the benchmark file [`/test/benchmark/map.js`](https://github.com/josdejong/mathjs/blob/develop/test/benchmark/map.js), I was triggered by an outlier `abs(array)` which is more than 10x faster than the rest:

```
# develop

abs(genericMatrix)                         x 54,106 ops/sec ±0.59% (96 runs sampled)
abs(array)                                 x 620,327 ops/sec ±0.51% (95 runs sampled)
abs(numberMatrix)                          x 48,865 ops/sec ±1.48% (92 runs sampled)
genericMatrix.map(abs)                     x 21,191 ops/sec ±0.62% (95 runs sampled)
numberMatrix.map(abs)                      x 19,951 ops/sec ±2.38% (94 runs sampled)
map(genericMatrix, abs)                    x 20,971 ops/sec ±0.56% (93 runs sampled)
map(numberMatrix, abs)                     x 20,036 ops/sec ±0.82% (93 runs sampled)
map(array, abs)                            x 21,609 ops/sec ±2.15% (93 runs sampled)
map(array, abs.signatures.number)          x 20,754 ops/sec ±2.60% (90 runs sampled)
genericMatrix.map(abs.signatures.number)   x 50,100 ops/sec ±1.08% (92 runs sampled)
numberMatrix.map(abs.signatures.number)    x 45,769 ops/sec ±0.94% (92 runs sampled)
```

The relevant code is the internal function `deepMap` and the method `DenseMatrix.map`:

https://github.com/josdejong/mathjs/blob/0f87a7b6c3456166cf355d8fc768a8fc9063847e/src/utils/collection.js#L56-L65

I think there is a bug in `deepMap` when executing on a Matrix: DenseMatrix already iterates over all items itself, so in that case there is no need to call deepMap recursively. This doesn't do much harm but it is more neat to change this to just call `array.map(callback)` in case of a Matrix, and only recurse in case of an Array.

After a little experiment I saw that the cause of the performance drop in `DenseMatrix.map` is caused by creating and passing an `index` array to every callback invocation. I think we can optimize this by checking if the callback actually uses `index`, and if not, do not provide it. I think we need to introduce an `recurseNoIndex` which is used when `callback.length === 1`.

I think we can best look into this after #3256 is merged since that PR refactors the functions `applyCallback` and `recurse` that needs futher changes. We also need to see how #3251 works out, since that also improves the `.map` function.

@dvd101x what do you think?

**Comment by kgryte** [comment]
Is `callback.length` a reliable indicator here? If this is only used with callbacks defined internally, then may be possible to get away with such a check, but, if users can provide callbacks, then they can also use `arguments` to access the index argument and relying on function arity may lead you astray.

**Comment by josdejong** [comment]
Good point. I think we can either pass a flag internally to tell whether or not indices are needed, or we could check something like `callback.toString().includes("arguments")` (a conservative check).

**Comment by dvd101x** [comment]
Hi Jos and Athan,

The implementation of `deepMap` is a bit different than other implementations, I wouldn't think it affects speed that much, but I will check if changing `if (array && (typeof array.map === 'function'))` to something else has some benefits.

I will do some tests and report back, maybe it's a simpler fix that could be implemented before #3256 or is something that can be fixed within #3256. It might take a few days.

## Maybe unrelated

Reading the comments it seems that the option to skip zeros in `deepMap` in some cases doesn't work.

https://github.com/josdejong/mathjs/blob/0f87a7b6c3456166cf355d8fc768a8fc9063847e/test/unit-tests/type/matrix/utils/deepMap.test.js#L56-L67

Currently there is some duplicated code with the various recurse functions and deepMap maybe when the performance issue is found, these different methods of iteration could be merged in a common place.

**Comment by dvd101x** [comment]
Hi Jos and Athan,

You were right on the money!

By doing both; fixing [`deepMap`](https://github.com/josdejong/mathjs/blob/develop/src/utils/collection.js#L56-L65) so that it doesn't later do [`DenseMatrix.map`](https://github.com/josdejong/mathjs/blob/develop/src/type/matrix/DenseMatrix.js#L537-L558) and adapting the recurse function so that it doesn't pass an index and an array regardless of the callback, some nice improvements were found.

```
abs(genericMatrix)                         x 107,296 ops/sec ±0.09% (99 runs sampled)
abs(array)                                 x 290,448 ops/sec ±0.46% (100 runs sampled)
abs(numberMatrix)                          x 85,916 ops/sec ±0.06% (100 runs sampled)
```
As a summary now `abs(genericMatrix)` is about 4 x faster than what it was. (in relation to `abs(array)`.

I made the following branch as a proof of concept.
https://github.com/josdejong/mathjs/tree/deepMap-and-recurse-performance

Please review as I think there are cleaner ways to know if the callback is called with 1, 2 or 3 arguments.

Jos, now I agree with you that this should be fixed after #3256 due to the refactors of `applyCallback` and `recurse` and the different nature of the change.

## Appendix 

I did some [tests in regular javascript](https://observablehq.com/@dvd101x/map-vs-for) and it seems that for regular functions (not typed) it doesn't affect that much in performance whether the callback is `callback(val)`, `callback(val, idx)` or `callback(val, idx, array)`.

**Comment by dvd101x** [comment]
After some tests, I think I found another opportunity to improve on performance besides the previously discussed.

Seeing how fast is `abs(array)` I think maybe recursion is not the biggest issue, at this point I'm not sure but I would like to work on this issue and make a PR.

# Proof of concept

For typed functions it would be nice to save the previously found signature implementation and try to use that first.

```js
const A = math.random([1000, 1000], -1, 1)
console.time('math.abs(Array)')
math.abs(A)
console.timeEnd('math.abs(Array)')
// yields 21.2 ms

console.time('math.map(Array, math.abs)')
math.map(A, math.abs)
console.timeEnd('math.map(Array, math.abs)')
// yields 435.9 ms

console.time('newMap(Array, math.abs)')
newMap(A, math.abs)
console.timeEnd('newMap(Array, math.abs)')
// yields 9.30 ms

function newMap(A, cb) {
  // make a dummy implementation that throws an error
  let implementation = () => {
    throw new Error('null implementation')
  }

  return _recurse(A)

  // recursive function that calls the implementation
  function _recurse(X) {
    if (Array.isArray(X)) return X.map(x => _recurse(x))
    try {
      // try to call the implementation
      return implementation(X)
    } catch {
      // resolve the implementation and call it
      implementation = math.typed.resolve(cb, [X]).implementation
      return implementation(X)
    }
  }
}
```

**Comment by dvd101x** [comment]
Hi,

Some progress regarding speed.

| name                                     | Master [ops/sec] | Fix DeepMap [ops/sec] |
| ---------------------------------------- | ---------------- | --------------------- |
| abs(genericMatrix)                       | 33,514           | 110,937               |
| abs(array)                               | 292,829          | 297,497               |
| abs(numberMatrix)                        | 31,285           | 88,499                |
| genericMatrix.map(abs)                   | 11,986           | 56,688                |
| numberMatrix.map(abs)                    | 11,601           | 50,054                |
| map(genericMatrix, abs)                  | 11,864           | 56,409                |
| map(numberMatrix, abs)                   | 11,844           | 50,208                |
| map(array, abs)                          | 12,278           | 85,195                |
| map(array, abs.signatures.number)        | 13,237           | 85,204                |
| genericMatrix.map(abs.signatures.number) | 32,086           | 105,399               |
| numberMatrix.map(abs.signatures.number)  | 25,089           | 85,757                |

I will try to make a PR with some improvements on the deepMap and other recursion functions following.

I have this running benchmark test that I'm using as a proof of concept. (Works best on chromium)
https://dvd101x.github.io/mathjs-fast-map/

**Comment by dvd101x** [comment]
Hi, the main idea is mostly done at.
https://github.com/josdejong/mathjs/tree/deepMap-and-recurse-performance

As a summary you right from the beginning since using the right amount of arguments specially when no index is needed, speeds up not only `deepMap` with matrix, but everything.

When a second argument `index` is included, a faster method was found. This could be applied in a separate PR.

The current develop branch, includes only one `recurse` algorithm that resembles `map`, if a separate methods are applied for recursion like `map` and `forEach` is made then also there is a speed improvement in `forEach` and `deepForEach`. This also could be applied in a separate PR.

Theoretically `map` and `forEach` could be faster than the current `abs(array)` according to https://github.com/dvd101x/mathjs-fast-map

<img width="653" alt="callback value" src="https://github.com/user-attachments/assets/15ed66c0-ecbe-4b19-a3ee-c941b506046f">

<img width="673" alt="callback valueIndex" src="https://github.com/user-attachments/assets/047f8d77-0631-47bc-a0d9-541badde15d6">

 I would like to know your opinions before making a PR.

**Comment by dvd101x** [comment]
Sorry for the frequent comments, I had some spare time and found the issue very intriguing.

I made PR #3266

I will chill for a bit.

**Comment by josdejong** [comment]
Awesome 😎

**Comment by dvd101x** [comment]
Thanks to #3251 this issue is reduced significantly. As you previously mentioned, now it's a matter of implementing it not only in matrices but also arrays and collections.

Even though the attempts to eliminate bottlenecks in recursion functions gave some nice theoretical results, by implementing them they didn't pan out as well as #3251.

**Comment by dvd101x** [comment]
Hi, with many PRs from the community, this issue has been improved overtime.

This discussion started at 13.1.1 which did have a regression in performance trying to fix an issue. Here I'm showing it in a heat map how this issue has evolved.

![Image](https://github.com/user-attachments/assets/11a097e3-f9b5-485f-9359-0cde7e077c29)
![Image](https://github.com/user-attachments/assets/eb9cf6eb-12ba-48f7-81c1-777c6570bf1c)

I think most of the low hanging fruit is already addressed, other than #3446 I think this topic is already addressed.

**Comment by josdejong** [comment]
Thanks for the overview David 😎 

Time to close this issue?

**Comment by dvd101x** [comment]
Thanks Jos 😁, yes I think it's ready to close. I don't see any immediate improvements to be made.

On the following weeks I will do a cleanup of some branches and incomplete PRs.

**Comment by josdejong** [comment]
👍 sounds good.
